Mathlib/Algebra/Ring/Ulift.lean

```diff
@@ -0,0 +1,195 @@
+  { zero := (0 : ULift α), add := (· + ·), mul := (· * ·), nsmul := AddMonoid.nsmul, add_assoc,
+    zero_add, add_zero, add_comm, left_distrib, right_distrib, zero_mul, mul_zero, mul_assoc,
+    mul_comm, nsmul_zero := fun _ => AddMonoid.nsmul_zero _,
+    nsmul_succ := fun _ _ => AddMonoid.nsmul_succ _ _ }
+#align ulift.non_unital_comm_semiring ULift.nonUnitalCommSemiring
+
+instance commSemiring [CommSemiring α] : CommSemiring (ULift α) :=
+  { ULift.semiring with
+      zero := (0 : ULift α), one := (1 : ULift α), add := (· + ·), mul := (· * ·),
+      nsmul := AddMonoid.nsmul, npow := Monoid.npow, mul_comm }
+#align ulift.comm_semiring ULift.commSemiring
+
+instance nonUnitalNonAssocRing [NonUnitalNonAssocRing α] : NonUnitalNonAssocRing (ULift α) :=
+  { zero := (0 : ULift α), add := (· + ·), mul := (· * ·), sub := Sub.sub, neg := Neg.neg,
+    nsmul := AddMonoid.nsmul, zsmul := SubNegMonoid.zsmul, add_assoc, zero_add, add_zero,
+    add_left_neg, add_comm, left_distrib, right_distrib, zero_mul, mul_zero, sub_eq_add_neg,
+    nsmul_zero := fun _ => AddMonoid.nsmul_zero _,
+    nsmul_succ := fun _ _ => AddMonoid.nsmul_succ _ _,
+    zsmul_zero' := SubNegMonoid.zsmul_zero', zsmul_succ' := SubNegMonoid.zsmul_succ',
+    zsmul_neg' := SubNegMonoid.zsmul_neg' }
+#align ulift.non_unital_non_assoc_ring ULift.nonUnitalNonAssocRing
+
+instance nonUnitalRing [NonUnitalRing α] : NonUnitalRing (ULift α) :=
+  { zero := (0 : ULift α), add := (· + ·), mul := (· * ·), sub := Sub.sub, neg := Neg.neg,
+    nsmul := AddMonoid.nsmul, zsmul := SubNegMonoid.zsmul, add_assoc, zero_add, add_zero, add_comm,
+    add_left_neg, left_distrib, right_distrib, zero_mul, mul_zero, mul_assoc, sub_eq_add_neg
+    nsmul_zero := fun _ => AddMonoid.nsmul_zero _,
+    nsmul_succ := fun _ _ => AddMonoid.nsmul_succ _ _,
+    zsmul_zero' := SubNegMonoid.zsmul_zero', zsmul_succ' := SubNegMonoid.zsmul_succ',
+    zsmul_neg' := SubNegMonoid.zsmul_neg' }
+#align ulift.non_unital_ring ULift.nonUnitalRing
+
+instance nonAssocRing [NonAssocRing α] : NonAssocRing (ULift α) :=
+  { zero := (0 : ULift α), one := (1 : ULift α), add := (· + ·), mul := (· * ·), sub := Sub.sub,
+    neg := Neg.neg, nsmul := AddMonoid.nsmul, zsmul := SubNegMonoid.zsmul,
+    intCast_ofNat := addGroupWithOne.intCast_ofNat, add_assoc, zero_add,
+    add_zero, add_left_neg, add_comm, left_distrib, right_distrib, zero_mul, mul_zero, one_mul,
+    mul_one, sub_eq_add_neg, nsmul_zero := fun _ => AddMonoid.nsmul_zero _,
+    nsmul_succ := fun _ _ => AddMonoid.nsmul_succ _ _,
+    zsmul_zero' := SubNegMonoid.zsmul_zero', zsmul_succ' := SubNegMonoid.zsmul_succ',
+    zsmul_neg' := SubNegMonoid.zsmul_neg',
+    natCast_zero := AddMonoidWithOne.natCast_zero, natCast_succ := AddMonoidWithOne.natCast_succ,
+    intCast_negSucc := AddGroupWithOne.intCast_negSucc }
+#align ulift.non_assoc_ring ULift.nonAssocRing
+
+-- Porting note: this doesn't work without `refine'`
```

Comment:
I don't understand why this one is different. It also looks pretty slow.

Reply:
It's to do with the definitions of `Int.cast`. The `Int.cast` field for this instance is filled in automatically as the recursive definition, but it should be inherited from the `Int.cast` instance on `R`. This should be changed and same with `Nat.cast` for this and the other instances.

Reply:
Adding `natCast` and `intCast` indeed fixed it, thanks!